py/dml/codegen.py

```diff
@@ -238,6 +245,135 @@ def codegen_exit(self, site, retvals):
         assert retvals is not None, 'dml 1.2/1.4 mixup'
         return codegen_return(site, self.outp, self.throws, retvals)
 
+def memoized_return_failure_leave(site, mkRef, failed):
+    ran = mkRef('ran', TInt(8, True))
+    threw = mkRef('threw', TBool())
+    stmts = [mkCopyData(site, mkIntegerLiteral(site, 1), ran),
+             mkCopyData(site, mkBoolConstant(site, failed), threw),
+             mkReturn(site, mkBoolConstant(site, failed))]
+    return stmts
+
+class MemoizedReturnFailure(Failure):
+    '''Generate boolean return statements to signal success. False
+    means success.'''
+
+    def __init__(self, site, mkRef):
+        self.site = site
+        self.mkRef = mkRef
+
+    def fail(self, site):
+        return mkCompound(
+            site,
+            memoized_return_failure_leave(site, self.mkRef, True))
+    def nofail(self):
+        '''Return code that is used to leave the method successfully'''
+        return mkCompound(
+            self.site,
+            memoized_return_failure_leave(self.site, self.mkRef, False))
+
+class MemoizedReturnExit(ExitHandler):
+    def __init__(self, site, outp, throws, mkRef):
+        self.site = site
+        self.outp = outp
+        self.throws = throws
+        self.mkRef = mkRef
+
+    def codegen_exit(self, site, retvals):
+        ran = self.mkRef('ran', TInt(8, True))
+        stmts = [mkCopyData(site, mkIntegerLiteral(site, 1), ran)]
+        if self.throws:
+            threw = self.mkRef('threw', TBool())
+            stmts.append(mkCopyData(site, mkBoolConstant(site, False), threw))
+        targets = []
+        for ((name, typ), val) in zip(self.outp, retvals):
+            target = self.mkRef(f'p_{name}', typ)
+            stmts.append(mkCopyData(site, val, target))
+            targets.append(target)
+        stmts.append(codegen_return(site, self.outp, self.throws, targets))
+        return mkCompound(site, stmts)
+
+class Memoization:
+    def prelude(self):
+        pass
+    def exit_handler(self):
+        pass
+    def fail_handler(self):
+        pass
+
+class IndependentMemoized(Memoization):
+    def __init__(self, func):
+        assert func.method.independent
+        self.func = func
+        self.method = func.method
+
+    def mkRef(self, ref, typ):
```

Comment:
PEP8 violation, albeit somewhat understandable.

Reply:
You'll need to clarify, I don't understand what specifically you're referring to.

Reply:
It was about dromedary case! Doing `mkRef` -> `make_ref` is simple enough.